File: packages/react-native-vision-camera-barcode-scanner/android/src/main/java/com/margelo/nitro/camera/barcodescanner/HybridBarcodeScanner.kt
package com.margelo.nitro.camera.barcodescanner

import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.common.InputImage
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.barcodescanner.extensions.toInputImage
import com.margelo.nitro.camera.barcodescanner.extensions.toMLBarcodeScannerOptions
import com.margelo.nitro.core.Promise
import com.margelo.nitro.image.HybridImageSpec
import java.util.concurrent.CancellationException

class HybridBarcodeScanner(
  options: BarcodeScannerOptions,
) : HybridBarcodeScannerSpec() {
  private val scanner = BarcodeScanning.getClient(options.toMLBarcodeScannerOptions())

  @OptIn(ExperimentalGetImage::class)
  override fun scanCodes(frame: HybridFrameSpec): Array<HybridBarcodeSpec> {
    val inputImage = frame.toInputImage()
    val task = scanner.process(inputImage)
    val barcodes = Tasks.await(task)
    return barcodes
      .map { HybridBarcode(it) }
      .toTypedArray<HybridBarcodeSpec>()
  }

  override fun scanCodesAsync(frame: HybridFrameSpec): Promise<Array<HybridBarcodeSpec>> {
    val inputImage = frame.toInputImage()
    return process(inputImage)
  }

  override fun scanCodesInImageAsync(image: HybridImageSpec): Promise<Array<HybridBarcodeSpec>> {
    val inputImage = image.toInputImage()
    return process(inputImage)
  }

  private fun process(inputImage: InputImage): Promise<Array<HybridBarcodeSpec>> {
    val promise = Promise<Array<HybridBarcodeSpec>>()

    scanner
      .process(inputImage)
      .addOnSuccessListener { barcodes ->
        val hybridBarcodes =
          try {
            barcodes
              .map { HybridBarcode(it) }
              .toTypedArray<HybridBarcodeSpec>()
          } catch (error: Throwable) {
            promise.reject(error)
            return@addOnSuccessListener
          }
        promise.resolve(hybridBarcodes)
      }.addOnFailureListener { error ->
        promise.reject(error)
      }.addOnCanceledListener {
        promise.reject(CancellationException("Barcode scan was cancelled."))
      }

    return promise
  }

  override fun dispose() {
    super.dispose()
    scanner.close()
  }
}

File: packages/react-native-vision-camera-barcode-scanner/ios/HybridBarcodeScanner.swift
//
//  HybridBarcodeScanner.swift
//  VisionCameraBarcodeScanner
//
//  Created by Marc Rousavy on 08.02.26.
//

import MLKitBarcodeScanning
import MLKitVision
import NitroImage
import NitroModules
import VisionCamera

class HybridBarcodeScanner: HybridBarcodeScannerSpec {
  private let scanner: BarcodeScanner

  init(options: BarcodeScannerOptions) {
    self.scanner = BarcodeScanner.barcodeScanner(options: options.toMLKitOptions())
    super.init()
  }

  func scanCodes(frame: any HybridFrameSpec) throws -> [any HybridBarcodeSpec] {
    let mlImage = try frame.toMLImage()
    let pointTransform = Self.getBufferToFrameTransform(for: frame)
    let barcodes = try scanner.results(in: mlImage)
    return barcodes.map { HybridBarcode(barcode: $0, pointTransform: pointTransform) }
  }

  func scanCodesAsync(frame: any HybridFrameSpec) throws -> Promise<[any HybridBarcodeSpec]> {
    let mlImage = try frame.toMLImage()
    let pointTransform = Self.getBufferToFrameTransform(for: frame)
    return process(mlImage, pointTransform: pointTransform)
  }

  func scanCodesInImageAsync(image: any HybridImageSpec) throws -> Promise<[any HybridBarcodeSpec]> {
    let mlImage = try image.toMLImage()
    return process(mlImage)
  }

  /**
   * MLKit on iOS returns coordinates in the raw pixel-buffer's coordinate
   * space - `MLImage.orientation` is only a detection hint. VisionCamera's
   * Frame coordinates are in oriented (display) space, so rotate/mirror the
   * MLKit results to match.
   */
  private static func getBufferToFrameTransform(for frame: any HybridFrameSpec)
    -> CGAffineTransform
  {
    let bufferWidth = frame.width
    let bufferHeight = frame.height

    // 1. Rotate buffer coordinates into upright (oriented) coordinates.
    //    `orientation` means: upright = buffer rotated counter-clockwise by
    //    `orientation.degrees`.
    var transform: CGAffineTransform
    let orientedWidth: Double
    switch frame.orientation {
    case .up:
      transform = .identity
      orientedWidth = bufferWidth
    case .left:
      // (x, y) -> (height - y, x)
      transform = CGAffineTransform(a: 0, b: 1, c: -1, d: 0, tx: bufferHeight, ty: 0)
      orientedWidth = bufferHeight
    case .right:
      // (x, y) -> (y, width - x)
      transform = CGAffineTransform(a: 0, b: -1, c: 1, d: 0, tx: 0, ty: bufferWidth)
      orientedWidth = bufferHeight
    case .down:
      // (x, y) -> (width - x, height - y)
      transform = CGAffineTransform(a: -1, b: 0, c: 0, d: -1, tx: bufferWidth, ty: bufferHeight)
      orientedWidth = bufferWidth
    }

    // 2. Mirror the upright image by flipping its x-axis.
    if frame.isMirrored {
      let mirror = CGAffineTransform(a: -1, b: 0, c: 0, d: 1, tx: orientedWidth, ty: 0)
      transform = transform.concatenating(mirror)
    }

    return transform
  }

  private func process(
    _ image: MLImage,
    pointTransform: CGAffineTransform = .identity
  ) -> Promise<[any HybridBarcodeSpec]> {
    let promise = Promise<[any HybridBarcodeSpec]>()

    scanner.process(image) { barcodes, error in
      if let error {
        promise.reject(withError: error)
        return
      }
      if let barcodes {
        promise.resolve(
          withResult: barcodes.map { HybridBarcode(barcode: $0, pointTransform: pointTransform) })
      } else {
        promise.resolve(withResult: [])
      }
    }

    return promise
  }
}
